File: 2022/C/01/part2.c
#include <assert.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#define MAX_SIZE 4096
#define TEST_INPUT_FILENAME "test_input.txt"
#define TEST_INPUT_ANSWER 45000

void bubble_sort(long* maxes) {
  long temp;
  bool sorted = false;
  while (!sorted) {
    sorted = true;
    for (int i = 1; i < 4; ++i) {
      if (maxes[i] > maxes[i - 1]) {
        sorted = false;
        temp = maxes[i - 1];
        maxes[i - 1 ] = maxes[i];
        maxes[i] = temp;
      }
    }
  }
}

int main(int argc, char** argv) {
  if (argc < 2) {
    fprintf(stderr, "Usage: part2 <filename>\n");
    return 1;
  }

  FILE* fd = fopen(argv[1], "r");
  if (fd == NULL) {
    perror("Error");
    return 1;
  }

  char* input = malloc(MAX_SIZE);
  long total = 0;
  long maxes[4] = {0};

  while(fgets(input, MAX_SIZE, fd)) {
    if (strcmp(input, "\n") != 0) {
      total += atoi(input);
    } else {
      maxes[3] = total;
      bubble_sort(maxes);
      total = 0;
    }
  }

  // Must do this as sometimes the file does not contain a blank extra line
  maxes[3] = total;
  bubble_sort(maxes);

  printf("Max: %li\n", maxes[0] + maxes[1] + maxes[2]);

  free(input);
  fclose(fd);

  if (strcmp(argv[1], TEST_INPUT_FILENAME) == 0 ) {
    assert((maxes[0] + maxes[1] + maxes[2]) == TEST_INPUT_ANSWER);
    puts("Test Passed!");
  }

  return 0;
}

File: 2022/Java/01/Part2.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;

public class Part2 {
  private static void bubble_sort(int maxes[]) {
    return;
  }
  
  public static void main(String args[]) {
    try {

      if (args.length < 1) {
        System.err.println("Usage: part1 <filename>");
        System.exit(1);
      }

      String filename = args[0];
      File f = new File(filename);
      Scanner fileReader = new Scanner(f);
      int lineTotal = 0;
      int max[] = new int[4];

      while(fileReader.hasNext()) {
        try {
          // If parsing was correct then add to line total;
          int line = Integer.parseInt(fileReader.nextLine());
          lineTotal += line;
        } catch (NumberFormatException nfe) {
          max[3] = lineTotal;
          bubble_sort(max);
          lineTotal = 0;
        }
      }

      fileReader.close();

      System.out.printf("Max Total: %d\n", (max[0] + max[1] + max[2]));
      if (filename.equals("test_input.txt")) {
        assert (max[0] + max[1] + max[2]) == 45000 : "Incorrect Answer";
        System.out.println("Test Passed!");
      }
    } catch (FileNotFoundException fnfe) {
      System.out.printf("Error: %s\n", fnfe.getMessage());
    } catch (Exception e) {
      System.out.printf("There was a problem: %s\n", e.getMessage());
    }
  }
}
